docs/specification.md

```diff
@@ -0,0 +1,555 @@
+- `i_u` determines the highest tick index in which this position will be active.
+- `i_l_l` is a witness (already initialized tick) index lower than `i_l`.
+  It should be as close as possible to `i_l`, for efficiency.
+- `i_u_l` is a witness (already initialized tick) index lower than `i_u`.
+  It should be as close as possible to `i_u`, for efficiency.
+- The liquidity of the `SENDER` will be updated by `delta_liquidity`, increased
+  or decreased depending on the sign, for both tokens of the pair.
+- In case, after adding eventual accrued fees, the `delta_liquidity` is:
+  * positive in `X` and/or `Y`: this amount will be `transfer`red **from** the `SENDER`
+  * negative in `X`: this amount will be `transfer`red **to** `to_x`
+  * negative in `Y`: this amount will be `transfer`red **to** `to_y`
+- If the position update is no longer acceptable because the `deadline` was not
+  met, fails with `TIMED_OUT`
+- If the amount of tokens that needs to be `transfer`red to the contract is
+  higher than `maximum_tokens_contributed`, fails with `HIGH_TOKENS`.
+
+```ocaml
+type set_position_param = {
+    i_l : tick_index ;
+    i_u : tick_index ;
+    i_l_l : tick_index ;
+    i_u_l : tick_index ;
+    delta_liquidity : int ;
+    to_x : address ;
+    to_y : address ;
+    deadline : timestamp ;
+    maximum_tokens_contributed : nat ;
+}
+```
+
+### **snapshot_cumulatives_inside**
+
+Oracle `view` for a snapshot of the tick cumulative, seconds per liquidity and
+seconds inside the given range.
+
+- `i_l` determines the lowest tick index of the range.
+- `i_u` determines the highest tick index of the range.
+- the `callback` contract will be called with the computed values.
+- `tick_cumulative_inside` is the computed snapshot of the tick cumulative for
+  the given range.
+- `seconds_per_liquidity_inside` is the computed snapshot of the seconds per
+  liquidity for the given range.
+- `seconds_inside` is the computed snapshot of the seconds for the given range.
+
+```ocaml
+type sci_view_param = {
+    tick_cumulative_inside : int ;
+    seconds_per_liquidity_inside : nat ;
+    seconds_inside : nat ;
+}
+
+type sci_param = {
+    i_l : tick_index ;
+    i_u : tick_index ;
+    callback : sci_view_param contract ;
+}
+```
+
+
+### **observe**
+
+Oracle `view` for the cumulative tick and liquidity-in-range, calculated for each
+`timestamp`s in a list.
+
+- Each value in `times` will be used to calculate a `cumulative_entry`.
+  The order of the computed entries is the same as that of the given `timestamp`s.
+- The `callback` contract will be called with the computed values.
+- For each `timestamp`/`cumulative_entry` pair:
+  * `tick_cumulative` is the cumulative tick value at that `timestamp`.
+  * `seconds_per_liquidity_cumulative` is the cumulative seconds per
+    liquidity-in-range at that `timestamp`.
+- If any of the given timestamps is too far back in the past,
+  fails with `INVALID_TIMESTAMP`.
```

Comment:
I think it's worth mentioning the motivation for storing a finite number of observations right here. That is a reasonable question that will probably arise in the mind of anyone who's reading this point.

AFAIU we were asked to avoid `big_map`s because they are costly, and generally adding some info to storage after each update in the contract is expensive for Tezos nodes?

Reply:
Done